### ISSUE TITLE
Acceptance criteria

### ISSUE THREAD

**Issue by gburgett** [rileyjhardy/docker_media_app#1]
Hey Riley!  Great Work!  I'm going to copy-paste the acceptance criteria we sent you here and check them off as I evaluate them to make sure your application does everything that we asked for.  Don't worry if I haven't checked everything off yet, I'll let you know when I'm finished going through the app and we can talk through your solution.

Acceptance Criteria
- [x] 1. The git repository SHALL contain a Ruby on Rails application
- [x] 2. The git repository SHALL contain a Readme file which describes:
  - [x] a. How to install the necessary code dependencies (e.g., “bundle install” or “docker-compose build”
  - [x] b. How to start the application including all dependent services (e.g., postgres, MySQL)
  - [x] c. What web address to navigate to in order to use the application
  - [x] d. (optional) How to run the test suite
- [x] 3. The application SHALL use a SQL database (Postgres, MySQL, or SQLite are acceptable)
- [x] 4. The application SHALL be accessible via a web browser after starting
- [x] 5. The homepage (path `/`) SHALL present a list of messages from the database
- [x] 6. The homepage SHALL provide a link to a form to upload a new message
- [x] 7. The new message page SHALL accept a mp4 file upload
- [x] 8. The new message page MAY accept an additional jpeg or PNG file upload as a thumbnail image
- [x] 9. The new message page SHALL accept the following textual inputs:
  - [x] a. Title
  - [x] b. Description c. Tag
- [x] 10. The new message page SHALL have a Submit button which creates the new message.
- [x] 11. The new message page SHALL have a Cancel button which discards the new message.
- [x] 12. The new message page SHALL validate that title and tag exist, rejecting the submission if validation fails.
- [x] 13. On validation failure, the new message page SHALL display an appropriate error message.
- [x] 14. On validation success, the new message page SHALL redirect to the homepage.
- [x] 15. The homepage SHALL display the newly created message in the list upon successful submission
- [x] 16. The homepage MAY display the thumbnail image alongside the newly created message as an HTML img tag
- [x] 17. The homepage SHALL provide a link to edit the message
- [x] 18. The edit message page SHALL accept the following textual inputs:
  - [x] a. Title
  - [x] b. Description c. Tag
- [x] 19. The edit message page MAY allow replacing the thumbnail image or mp4 video file
- [x] 20. The edit message page SHALL have a Submit button which creates the new message.
- [x] 21. The edit message page SHALL have a Cancel button which discards the in-progress edit.
- [x] 22. The edit message page SHALL validate that title and tag exist, rejecting the submission if validation fails.
- [x] 23. On validation failure, the edit message page SHALL display an appropriate error message.
- [x] 24. On validation success, the edit message page SHALL redirect to the homepage
- [x] 25. All HTML pages in the application SHOULD have a consistent style and use a grid system (e.g., Bootstrap, Flexbox, CSS grid)
- [x] 26. The application SHALL expose an API endpoint to query messages from the database
- [x] 27. The API query endpoint SHALL respond with the `application/json` content type
- [x] 28. The API query endpoint SHALL accept the `?tag=` HTTP query parameter to filtermessages by tag
- [x] 29. When filtering messages by tag, the API query endpoint SHALL only include messages matching the provided tag
- [x] 30. The API query endpoint’s response SHALL include the following in JSON format for each message:
  - [x] * Title
  - [x] * Description * Tag
  - [x] * URL
- [x] 31. The above URL SHALL cause the browser to play the uploaded MP4 video when pasted into the URL bar.
- [x] 32. The API query endpoint’s response MAY include the following additional fields: * thumbnail image URL
- [x] 33. The above thumbnail image URL, if included, SHALL result in an image being displayed when used as the “src” attribute in an HTML image element
- [x] 34. The application MAY contain a suite of unit tests, and instructions on how to execute them.

**Comment by gburgett** [comment]
Riley this is fantastic work!  I'm very surprised at how quickly you were able to figure all this out.

There's only one thing you missed, your response doesn't provide the full URL to the image or the media file.  I can see that the application does serve the media file via the relative URL, but without the protocol, host, and port of the URL I can't paste it into the URL bar and go there automatically.  This kind of thing is important because downstream apps consuming this API will need to know which host and port to go to.

Other than that, great job on doing almost everything the "Rails" way!

Some non-critical feedback that doesn't affect the evaluation of the project, but is intended to help you improve in the future:

1. Ideally if you're using docker, the only instructions that we should need in order to get the thing running are "step 1: run `docker-compose up`, step 2: navigate to localhost:3000`.  Your other steps of `bundle install` and `db migrate` should be handled in docker build if you're bundling the sqlite database, or in a startup script if you're not bundling the database.
2. If I `docker-compose down` and then `docker-compose up` again, I lose the database and the uploaded files.  I also have to bundle install again.  Your compose file should declare some volumes so that these are persisted on disk.
3. Your specs inside `spec/models` aren't verifying anything really related to your application code, they're mostly verifying just default Rails behaviors.  What you really want to do is exercise the controller logic for creating and updating messages, because you're more likely to find a bug if you exercise that whole stack than if you just exercise the default rails behaviors.  But that's an experience call that comes with time!

Again, nothing of the above reflects poorly on you!  Very impressive work.

**Comment by rileyjhardy** [comment]
Thank you @gburgett ! Rails (and Ruby) is very different than the stuff I've been learning, but I've really enjoyed learning about the framework!

That explanation makes a lot of sense about the URL. I knew it was a relative URL, but I wasn't sure what the intended use case was going to be.

Thank you for the feedback! That's super helpful. Testing is a new topic for me so that was definitely the most difficult part. There were a few other request specs I wanted to add, but I couldn't get them to work haha.

**Comment by rileyjhardy** [comment]
@gburgett I made some changes to include the protocol, host, and port! Would you say a carrierwave initializer is a good way to go about it?

**Comment by gburgett** [comment]
Yep I think that's exactly the right way to do it!  I've actually never used CarrierWave, though it seems like the right solution for this kind of thing going forward.  We have a legacy system that we run called "Tessa" which manages uploads for us.  It was the right choice when we needed it 5+ years ago, but now there's some open source alternatives that would be better.  We just have never felt a need to do the work to convert over to CarrierWave.
